aiomisc/aggregate.py

```diff
@@ -0,0 +1,144 @@
+import asyncio
+import inspect
+from asyncio import CancelledError, Event, Future, Lock, wait_for
+from contextlib import suppress
+from inspect import Parameter
+from time import monotonic
+from typing import Any, Awaitable, Callable, Iterable, List, Optional
+
+AggFunc = Callable[[Any], Awaitable[Iterable]]
+
+
+class Aggregator:
+
+    def __init__(
+            self, func: AggFunc, *,
+            leeway_ms: float, max_count: int = None,
+    ):
+        has_variadic_positional = any((
+            parameter.kind == Parameter.VAR_POSITIONAL
+            for parameter in inspect.signature(func).parameters.values()
+        ))
+        if not has_variadic_positional:
+            raise ValueError(
+                "Function must accept variadic positional arguments",
+            )
+
+        if max_count is not None and max_count <= 0:
+            raise ValueError("max_count must be positive int or None")
+
+        if leeway_ms <= 0:
+            raise ValueError("leeway_ms must be positive float")
+
+        self._func = func
+        self._max_count = max_count
+        self._leeway = leeway_ms / 1000
+        self._t = None          # type: Optional[float]
+        self._args = []         # type: list
+        self._futures = []      # type: List[Future]
+        self._event = Event()   # type: Event
+        self._lock = Lock()     # type: Lock
+
+    @property
+    def max_count(self) -> Optional[int]:
+        return self._max_count
+
+    @property
+    def leeway_ms(self) -> float:
+        return self._leeway * 1000
+
+    @property
+    def count(self) -> int:
+        return len(self._args)
+
+    def _clear(self) -> None:
```

Comment:
May be call in `__init__` instead set same vars in init?  One day some one can add another one var in one place and forget to add in other.

Reply:
\+